networks/suzuka/suzuka-full-node/src/partial.rs

```diff
@@ -138,6 +138,14 @@ where
             ).await?
         }.into_inner();
 
+        let settlement_client = self.settlement_client.clone();
+        // TODO: consume the commitment stream to finalize blocks
+        let commitment_stream = settlement_client.stream_block_commitments().await?;
+
+        tokio::spawn(async move {
+            process_commitments(receiver, settlement_client).await;
+        });
```

Comment:
This is less well organized than tasks-as-methods, but it enables parallel execution. `try_join!` only gives you a compound future that's polled by a single thread.

Unfortunately, the top-down design dictated by traits like `SuzukaFullNode` where the "god object" is borrowed immutably as `&self` by async methods does not work well with background tasks that ideally need to work on moved, rather than shared, bits of state. You end up needing `Arc<Mutex<_>>` for anything mutable even if conceptually it's only used in one task. Another smell is needing `async_channel`: there is hardly ever a genuine need to have multiple consumers on the channel, you are only forced to clone it by the constraints on higher-level API, but it's a slower, more synchronization-intensive channel implementation than `tokio::sync::mpsc`.

Reply:
Which tasks specifically do you want to have work on moved instead of shared state?

Reply:
Operations on receivers I take it?

Reply:
I think the easiest workaround is just to have all methods using the receiver be `self` which is fairly straightforward refactor from here. Maybe toss into a new issue?

Reply:
Yes, node object's public API should perhaps have only two stages:

- Synchronous construction, either with a simple associated constructor fn, or, if needed, a builder API;
- Something like `async fn run(self) -> Result<(), NodeError>` to consume and run the instance. Maybe even `Result<!, NodeError>` to signal that the function never returns in the normal case.

Reply:
I have reworked this to be more along the lines suggested in https://github.com/movementlabsxyz/movement/issues/63#issuecomment-2120184646

The commitment processing loop has been moved out to the mcr-settlement-manager crate.